src/hotspot/cpu/x86/macroAssembler_x86.cpp

```diff
@@ -5869,15 +5869,31 @@ void MacroAssembler::generate_fill(BasicType t, bool aligned,
     BIND(L_skip_align2);
   }
   {
-    Label L_fill_32_bytes;
+    Label L_fill_32_bytes, L_C_align, L_align_64_bytes;
     if (!UseUnalignedLoadStores) {
       // align to 8 bytes, we know we are 4 byte aligned to start
       testptr(to, 4);
+      jccb(Assembler::zero, L_C_align);
+      movl(Address(to, 0), value);
+      addptr(to, 4);
+      subptr(count, 1<<shift);
+    }
+
+    BIND(L_C_align);
+    if (EnableX86ECoreOpts) {
+      // align 'big' arrays to 64 bytes (cache line size) to minimize split_stores
+      cmpptr(count, 256<<shift);
+      jcc(Assembler::below, L_fill_32_bytes);
+
```

Comment:
I see you have an overhead for small sizes, may be we could do a check for small sizes before line 5885 something like below:

movdl(xtmp, value);
vpbroadcastd(xtmp, xtmp, Assembler::AVX_256bit);
subptr(count, 16 << shift);
jcc(Assembler::less, L_check_fill_32_bytes);

Or alternatively move the entire if (EnableX86ECoreOpts) { } block of code to line 5933 adjusting the jump labels accordingly.